File: src/main/java/net/gtaun/shoebill/common/timers/TemporaryTimer.java
package net.gtaun.shoebill.common.timers;

import net.gtaun.shoebill.object.Timer;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * TemporaryTimer is used to create temporary timers than run for a short period of time.
 */
public class TemporaryTimer implements Timer {

    private static Timer recycleTimer;
    private static List<TemporaryTimer> timers;

    static {
        timers = new ArrayList<>();

        recycleTimer = Timer.create(5000, i -> {
            Iterator<TemporaryTimer> iterator = timers.iterator();
            while (iterator.hasNext()) {
                TemporaryTimer timer = iterator.next();
                if (!timer.isRunning() && timer.hasBeenStarted()) {
                    timer.destroy();
                    iterator.remove();
                }
            }
        });
        recycleTimer.start();
    }

    /**
     * Creates a temporary timer.
     *
     * @param interval In which interval the callback should be called.
     * @param callback The callback that will be executed when the interval has been reached.
     * @return The created TemporaryTimer instance.
     */
    public static TemporaryTimer create(int interval, TimerCallback callback) {
        return create(interval, Timer.COUNT_INFINITE, callback);
    }

    /**
     * Creates a temporary timer
     *
     * @param interval In which interval the callback should be called.
     * @param count    How often the callback should be executed before the timer is destroyed.
     * @param callback The callback that will be executed when the interval has been reached.
     * @return The created TemporaryTimer instance.
     */
    public static TemporaryTimer create(int interval, int count, TimerCallback callback) {
        return new TemporaryTimer(interval, count, callback);
    }

    private Timer timer;
    private boolean hasBeenStarted = false;

    private TemporaryTimer(int interval, int count, TimerCallback callback) {
        timer = Timer.create(interval, count, callback);
        timers.add(this);
    }

    /**
     * Gets the interval
     *
     * @return Interval
     */
    @Override
    public int getInterval() {
        return timer.getInterval();
    }

    /**
     * Gets the amount of TimerCallback calls.
     *
     * @return TimerCallback counts.
     */
    @Override
    public int getCount() {
        return timer.getCount();
    }

    /**
     * Returns the running state of the temporary timer.
     *
     * @return Running State
     */
    @Override
    public boolean isRunning() {
        return timer.isRunning();
    }

    /**
     * Sets the interval of the timer.
     *
     * @param ms Interval in milliseconds.
     */
    @Override
    public void setInterval(int ms) {
        timer.setInterval(ms);
    }

    /**
     * Sets the amount of TimerCallback counts.
     *
     * @param count Amount of calls.
     */
    @Override
    public void setCount(int count) {
        timer.setCount(count);
    }

    /**
     * Starts the temporary timer.
     */
    @Override
    public void start() {
        timer.start();
        hasBeenStarted = true;
    }

    /**
     * Stops the temporary timer (this might destroy the timer).
     */
    @Override
    public void stop() {
        timer.stop();
    }

    /**
     * Sets the callback of the temporary timer.
     *
     * @param callback The new TimerCallback.
     */
    @Override
    public void setCallback(TimerCallback callback) {
        timer.setCallback(callback);
    }

    /**
     * Destroys the temporary timer.
     */
    @Override
    public void destroy() {
        timer.destroy();
    }

    /**
     * Returns the state of the timer.
     *
     * @return State of Timer.
     */
    @Override
    public boolean isDestroyed() {
        return timer.isDestroyed();
    }

    private boolean hasBeenStarted() {
        return hasBeenStarted;
    }

}

File: src/main/java/net/gtaun/shoebill/common/timers/TemporaryTimer.kt
package net.gtaun.shoebill.common.timers

import net.gtaun.shoebill.constant.Collectable
import net.gtaun.shoebill.entities.Timer
import net.gtaun.shoebill.entities.TimerCallback

/**
 * TemporaryTimer is used to create temporary timers than run for a short period of time.
 */

class TemporaryTimer @JvmOverloads constructor(
        interval: Int,
        count: Int = Timer.COUNT_INFINITE,
        callback: TimerCallback? = null) : Timer() {

    private val timer: Timer = Timer.create(interval, count, callback)
    private var hasBeenStarted = false

    init {
        timers.add(this)
    }

    /**
     * The interval of the timer in milliseconds.
     */
    override var interval: Int
        get() = timer.interval
        set(ms) {
            timer.interval = ms
        }

    /**
     * The amount of TimerCallback calls.
     */
    override var count: Int
        get() = timer.count
        set(count) {
            timer.count = count
        }

    /**
     * The running state of the temporary timer.
     */
    override val isRunning: Boolean
        get() = timer.isRunning

    /**
     * Starts the temporary timer.
     */
    override fun start() {
        timer.start()
        hasBeenStarted = true
    }

    /**
     * Stops the temporary timer (this might destroy the timer).
     */
    override fun stop() = timer.stop()

    /**
     * Destroys the temporary timer.
     */
    override fun destroy() = timer.destroy()

    /**
     * The state of the timer.
     */
    override val isDestroyed: Boolean
        get() = timer.isDestroyed

    /**
     * The callback of the temporary timer.
     */
    override var callback: TimerCallback?
        get() = timer.callback
        set(callback) {
            timer.callback = callback
        }

    companion object : Collectable<TemporaryTimer> {

        /**
         * Creates a [TemporaryTimer] with params.
         */
        @JvmStatic
        fun create(interval: Int, callback: TimerCallback? = null): TemporaryTimer = TemporaryTimer.create(interval,
                COUNT_INFINITE, callback)

        /**
         * Creates a [TemporaryTimer] with params.
         * @param interval The interval in milliseconds.
         * @param count How often the Timer will get called.
         * @param callback The callback which will get invoked after the interval.
         * @return The created [TemporaryTimer].
         */
        @JvmStatic
        @JvmOverloads
        fun create(interval: Int, count: Int = COUNT_INFINITE, callback: TimerCallback? = null): TemporaryTimer =
                TemporaryTimer(interval, count, callback)

        private var recycleTimer: Timer = Timer.create(5000, Timer.COUNT_INFINITE, TimerCallback {
            val iterator = timers.iterator()
            while (iterator.hasNext()) {
                val timer = iterator.next()
                if (!timer.isRunning && timer.hasBeenStarted) {
                    timer.destroy()
                    iterator.remove()
                }
            }
        })

        private var timers: MutableList<TemporaryTimer> = mutableListOf()

        /**
         * Gets all available [TemporaryTimer] instances.
         */
        override fun get(): Collection<TemporaryTimer> = timers

        init {
            recycleTimer.start()
        }
    }
}